### PR TITLE
Add post about my benchmarking comparison of Rust, Kotlin, and Python

### PR DESCRIPTION
**Pull request by JoelEager**
This post compares the execution speed of a collision detection algorithm in Rust, Kotlin, and Python. The code discussed in it comes from my major projects from both this year and last year.

@devinmatte just as a heads up, the table formatting looked a bit off when I build this locally. You may want to modify the table CSS before this goes up so it's more readable. Thanks!

### CODE REVIEW

**Comment by RamZallan** [comment]
Content LGTM! Someone else may want to check the table formatting, since I can't build locally right now.

File: _posts/2018-03-26-benchmarking-programing-languages.md [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+the programming language being used.
+
+To explore this relationship I implemented a common collision detection algorithm called [separating axis theorem](https://gamedevelopment.tutsplus.com/tutorials/collision-detection-using-the-separating-axis-theorem--gamedev-169) 
+in Python, Rust, and Kotlin. This algorithm uses linear algebra to check if two convex 2D polygons are overlapping. 
+Since this is a linear algebra algorithm most of the work is repetitive arithmetic. This makes it a relatively good 
+test of raw execution speed.
+
+Even before running the code I can expect [Python](https://en.wikipedia.org/wiki/Python_(programming_language)) to be 
+the slowest since it's the only interpreted language being tested. [Rust](https://en.wikipedia.org/wiki/Rust_(programming_language)), 
+on the other hand, is complied down to machine code so it should be significantly faster. [Kotlin](https://en.wikipedia.org/wiki/Kotlin_(programming_language))
```
Comment (csssuf) [inline _posts/2018-03-26-benchmarking-programing-languages.md]: `s/complied/compiled/`

File: _posts/2018-03-26-benchmarking-programing-languages.md [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+To explore this relationship I implemented a common collision detection algorithm called [separating axis theorem](https://gamedevelopment.tutsplus.com/tutorials/collision-detection-using-the-separating-axis-theorem--gamedev-169) 
+in Python, Rust, and Kotlin. This algorithm uses linear algebra to check if two convex 2D polygons are overlapping. 
+Since this is a linear algebra algorithm most of the work is repetitive arithmetic. This makes it a relatively good 
+test of raw execution speed.
+
+Even before running the code I can expect [Python](https://en.wikipedia.org/wiki/Python_(programming_language)) to be 
+the slowest since it's the only interpreted language being tested. [Rust](https://en.wikipedia.org/wiki/Rust_(programming_language)), 
+on the other hand, is complied down to machine code so it should be significantly faster. [Kotlin](https://en.wikipedia.org/wiki/Kotlin_(programming_language)) 
+should be somewhere in the middle since its compiled but runs in the Java Virtual Machine instead of directly on the
```
Comment (csssuf) [inline _posts/2018-03-26-benchmarking-programing-languages.md]: `s/its/it's/`

File: _posts/2018-03-26-benchmarking-programing-languages.md [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+Since this is a linear algebra algorithm most of the work is repetitive arithmetic. This makes it a relatively good 
+test of raw execution speed.
+
+Even before running the code I can expect [Python](https://en.wikipedia.org/wiki/Python_(programming_language)) to be 
+the slowest since it's the only interpreted language being tested. [Rust](https://en.wikipedia.org/wiki/Rust_(programming_language)), 
+on the other hand, is complied down to machine code so it should be significantly faster. [Kotlin](https://en.wikipedia.org/wiki/Kotlin_(programming_language)) 
+should be somewhere in the middle since its compiled but runs in the Java Virtual Machine instead of directly on the 
+hardware.
+
+The metric I'll be measuring here is the wall-clock execution time. Note that this approach has a inherit amount of
```
Comment (csssuf) [inline _posts/2018-03-26-benchmarking-programing-languages.md]: `s/a inherit/an inherent/`

File: _posts/2018-03-26-benchmarking-programing-languages.md [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+| --------------------- | --------------- | --------------- | ------------- |
+| Intel Core i7-8550U   | 21.800 seconds  | 0.308 seconds   | 0.230 seconds |
+| AMD A10-6800K         | 40.617 seconds  | 0.767 seconds   | 0.465 seconds |
+
+Now this is a pretty naive implementation. The two objects are 10 and 1 pixels across respectively so it takes a lot of 
+iterations before they collide. This algorithm takes the most time to run when the objects aren't touching so it would 
+be quite helpful if there was a way to skip over at least some of the iterations where they aren't touching.
+
+It turns out that it's trivial to calculate a rough estimate of the distance between the two objects by taking one point 
+from each and applying the Pythagorean theorem. Now, this would be a very a rough estimate since it doesn't take into
```
Comment (csssuf) [inline _posts/2018-03-26-benchmarking-programing-languages.md]: `s/a very a/a very/`

File: _posts/2018-03-26-benchmarking-programing-languages.md [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+| AMD A10-6800K         | 40.617 seconds  | 0.767 seconds   | 0.465 seconds |
+
+Now this is a pretty naive implementation. The two objects are 10 and 1 pixels across respectively so it takes a lot of 
+iterations before they collide. This algorithm takes the most time to run when the objects aren't touching so it would 
+be quite helpful if there was a way to skip over at least some of the iterations where they aren't touching.
+
+It turns out that it's trivial to calculate a rough estimate of the distance between the two objects by taking one point 
+from each and applying the Pythagorean theorem. Now, this would be a very a rough estimate since it doesn't take into 
+account the rotation or shape of the objects in question. However, if that rough separation estimate is greater than the 
+combined width of the two objects at their widest points then it can be assumed that that objects aren't touching. In
```
Comment (csssuf) [inline _posts/2018-03-26-benchmarking-programing-languages.md]: `s/that that/that the/`

File: _posts/2018-03-26-benchmarking-programing-languages.md [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+be quite helpful if there was a way to skip over at least some of the iterations where they aren't touching.
+
+It turns out that it's trivial to calculate a rough estimate of the distance between the two objects by taking one point 
+from each and applying the Pythagorean theorem. Now, this would be a very a rough estimate since it doesn't take into 
+account the rotation or shape of the objects in question. However, if that rough separation estimate is greater than the 
+combined width of the two objects at their widest points then it can be assumed that that objects aren't touching. In 
+all other cases the program would still have to run the collision detection algorithm as before to get an accurate 
+result.
+
+With this optimization in play the execution times drop significantly. Note that due the the inherit inaccuracy of the
```
Comment (csssuf) [inline _posts/2018-03-26-benchmarking-programing-languages.md]: `s/due the the inherit/due to the inherent/`

**Comment by JoelEager** [comment]
I've pushed a commit that fixes those typos. Thanks @csssuf for pointing them out.

**Comment by JoelEager** [comment]
I think it makes the most sense to hold off on merging this until the table CSS is fixed. Thanks for approving it though.

**Comment by csssuf** [comment]
Do we want to bump the date?

**Comment by devinmatte** [comment]
@JoelEager If you could push the date up to tomorrow, I'll be ready to merge

**Comment by JoelEager** [comment]
The date is updated and the tables look good. Thanks for everyone's help on this.

**Pull request merged**